[Merged by Bors] - feat(data/set/lattice): add lemmas set.Union_subtype and set.Union_of_singleton_coe

adomani opened the pull request:
Add one simp lemma, following a suggestion on the Zulip chat:

https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/image_bUnion

---
<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

adomani committed:
add simp lemma set.bUnion_of_singleton_of_coe

adomani committed:
Remove `set.` to please the linter

adomani committed:
add lemma suggested by Eric Wieser

adomani committed:
Merge branch 'set.bUnion_of_singleton_of_coe' of github.com:leanprover-community/mathlib into set.bUnion_of_singleton_of_coe

eric-wieser commented on src/data/set/lattice.lean:
```diff
@@ -358,6 +358,14 @@ theorem bUnion_union (s t : set α) (u : α → set β) :
 supr_union
 
+@[simp] lemma bUnion_of_singleton_of_coe (s : set α) :
+  (⋃ (i : s), {i} : set α) = s :=
```
I think this name is incorrect - `⋃ (i ∈ s)` is a `bUnion`, while `⋃ (i : s)` is just a `Union`. That probably indicates that this lemma belongs at a different point in the file too.

adomani replied:
I am only moderately aware of these subtleties, so I may be wrong about this.  In the file, there are a lot of lemmas where there is a separate Type, the Union ranges over the separate Type, and there is an explicit function from this other Type to set \a.  In this case, probably the separate Type would be s (viewed as a Type) and the function is the "singleton of coercion".  I am not sure that I can formulate a simpler lemma, though.

Do you have any suggestion?  Also, I browsed the file, but I could not find where this lemma would go, if not where I put it.

eric-wieser replied:
I think this lemma belongs right next to `Union_of_singleton`, and called `Union_of_singleton_coe`.

https://github.com/leanprover-community/mathlib/blob/06c452762dbdc4ef9fad23b4d6e6fa9ae481e3fe/src/data/set/lattice.lean#L545-L546

I'm not suggesting you change the lemma, but just that the word `bUnion` should not appear in its name.

adomani replied:
I had missed your reply: I just moved the lemma to where you suggest and renamed it!  Thanks!

eric-wieser commented on src/data/set/lattice.lean:
```diff
@@ -358,6 +358,14 @@ theorem bUnion_union (s t : set α) (u : α → set β) :
+
+lemma set.Union_subtype {α β : Type*} (s : set α) (f : α → set β) :
+  (⋃ (i : s), f i) = ⋃ (i ∈ s), f i :=
+(set.bUnion_eq_Union s $ λ x _, f x).symm
```
For other reviewers - is this a good `simp` lemma? My feeling is yes, because we have a lot of other simp lemmas about things of the form `⋃ (i ∈ s)` when `s` is some compound expression.

adomani committed:
lint

adomani committed:
following Eric's suggestion:
moved lemma bUnion_of_singleton_of_coe
and renamed it to Union_of_singleton_coe

bryangingechen commented on src/data/set/lattice.lean:
```diff
@@ -358,6 +358,10 @@ theorem bUnion_union (s t : set α) (u : α → set β) :
   (⋃ x ∈ s ∪ t, u x) = (⋃ x ∈ s, u x) ∪ (⋃ x ∈ t, u x) :=
 supr_union
 
+lemma Union_subtype {α β : Type*} (s : set α) (f : α → set β) :
```
```suggestion
@[simp] lemma Union_subtype {α β : Type*} (s : set α) (f : α → set β) :
```
Let's try Eric's suggestion. The linter seems happy enough with it (at least in this file).

If it builds, feel free to merge. Thanks!
bors d+

adomani committed:
Update src/data/set/lattice.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>
